File: support/my_boot.py
# Vazeny bootstrap
#  index ... S*1 vektor bootstrapovych indexu
#  y ....... N x 1 vektor bootstrapovaneho vzorku
#  S ....... velikost bootstrapoveho vyberu
#  w ....... bootstrapove vahy (optional)

from support.progress_info import progress_bar
import numpy as np

def my_boot(y: np.ndarray, s: int, w=None):
    n = y.size
    # if not w:
    #     w = 1/n
    # Vektor indexu
    boot_index = [0] * s
    # Vytvoreni bodu intervalu, kde sirka dilcich intervalu reprezentuje stanovenou vahu
    w_int = np.cumsum(w)
    # Nastaveni seedu (muze byt libovolne cislo) - zakomentovano (neni nezbytne nutne)
    # np.random.seed(420)
    print("\nBootstrapping:")
    for i in range(s):
        threshold = np.random.rand() * w_int[-1]
        boot_index[i] = int(np.sum(w_int < threshold))   # konverze z float na int protoze se pozdeji pouziva jako index
        progress_bar(i, s)

    return boot_index

File: Témata/09 Tobit model/R/Support/my_boot.R
# Vazeny bootstrap
# index ... S*1 vektor bootstrapovych indexu
# y ... N x 1 vektor bootstrapovaneho vzorku
# S ... velikost bootstrapoveho vyberu
# w ... bootstrapove vahy (optional)

my_boot <- function(y,S,w = NULL) {
    N <- length(y)

    if(is.null(w)) {
        w <- 1/N
    }

    # vektor indexu
    boot_index <- mat.or.vec(S,1)

    # vytvoreni bodu intervalu, kde sirka dilcich intervalu reprezentuje
    # stanovenou vahu
    w_int <- cumsum(w)

    for(s in 1:S) {
        # runif*w_int(end) ...nahodne cislo z uniformniho rozdeleni na intervalu
        # (0;w_int(length(w)) a ziskani indexu jako suma intervalu splnujici
        # podminku, ze hodnota krajni meze intervalu je mensi nez toto nahodne
        # vygenerovane cislo
        boot_index[s] <- sum(w_int < (runif(1)*w_int[length(w)])) + 1
    }
    return(boot_index)
}
